src/training/communicator.h

```diff
@@ -84,6 +84,18 @@ struct/*interface*/ IMPIWrapper {
   virtual void sSend(void* buf, size_t count, MPI_Datatype datatype, size_t destRank, int tag, MPI_Comm comm = MPI_COMM_WORLD) const = 0;
   virtual void recv(void* buf, size_t count, MPI_Datatype datatype, size_t sourceRank, int tag, MPI_Comm comm = MPI_COMM_WORLD, MPI_Status* status = MPI_STATUS_IGNORE) const = 0;
   virtual void allReduce(const void* sendbuf, void* recvbuf, size_t count, MPI_Datatype datatype, MPI_Op op, MPI_Comm comm = MPI_COMM_WORLD) const = 0;
+  virtual void reduceScatter(const void * sendbuf, void * recvbuf, int * recvcounts, MPI_Datatype datatype, MPI_Op op, MPI_Comm comm = MPI_COMM_WORLD) const {
+    sendbuf; recvbuf; recvcounts; datatype; op; comm; // unused in the fakeMPI wrapper
+    ABORT("ReduceScatter is only implemented when compiled with -DUSE_MPI=ON");
+  }
```

Comment:
Since those are non-pure virtual functions with implementations, let's add empty lines around them for readability.

Reply:
Similar below.